jetcd-launcher/src/main/java/io/etcd/jetcd/launcher/EtcdClusterFactory.java

```diff
@@ -64,6 +78,9 @@ public void start() {
         } catch (InterruptedException e) {
           throw new RuntimeException(e);
         }
+        if (failedToStart.get()) {
+          throw new IllegalStateException("Cluster failed to start");
```

Comment:
I don't think we need this any more

Reply:
Looks like it's still needed. With this commented out `WatchResumeTest` fails fast but with following errors:
```
java.lang.IllegalArgumentException: Requested port (2379) is not mapped

	at org.testcontainers.containers.ContainerState.getMappedPort(ContainerState.java:116)
	at io.etcd.jetcd.launcher.EtcdContainer.clientEndpoint(EtcdContainer.java:186)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1492)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.etcd.jetcd.launcher.EtcdClusterFactory$2.getClientEndpoints(EtcdClusterFactory.java:98)
	at io.etcd.jetcd.WatchResumeTest.setUp(WatchResumeTest.java:52)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```
for `@BeforeClass` and
```
java.lang.NullPointerException
	at io.etcd.jetcd.WatchResumeTest.tearDown(WatchResumeTest.java:59)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

for `@AfterClass`

Without this line it looks like cluster was successfully created which isn't the case.

With this line the message is:
```
java.lang.IllegalStateException: Cluster failed to start

	at io.etcd.jetcd.launcher.EtcdClusterFactory$2.start(EtcdClusterFactory.java:82)
	at io.etcd.jetcd.launcher.junit.EtcdClusterResource.before(EtcdClusterResource.java:53)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:46)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

Reply:
There could be a bit of confusion regarding exceptions because each container starts in its own thread: https://github.com/etcd-io/jetcd/blob/f2f75dce0c0645464181735496e145727d4fffeb/jetcd-launcher/src/main/java/io/etcd/jetcd/launcher/EtcdClusterFactory.java#L59

So if there is an exception while starting container it won't be communicated back to main thread (one that creates `ExternalResource`). atomic boolean is needed to communicate that and `throw` is needed to signal that there's been an error and test cannot proceed with execution

Reply:
ah yeah sorry, got confused by github pr view. So I rather change the AtomicBoolean with AtomicReference and set it as cause of the new thrown exception

Reply:
Since we could have multiple nodes I thought it would be weird to store and re-throw only last exception. And if store all of them - which one to throw? So I've decided to just throw some exception to indicate that startup failed.
Do you want me to change it to throw the last captured exception?

Reply:
let merge it for the moment, we'll review it later.